File: Problems/No.2115_Find_All_Possible_Recipes_from_Given_Supplies/find_all_possible_recipes_from_given_supplies.cpp
class Solution {
public:
    vector<string> findAllRecipes(vector<string>& recipes, vector<vector<string>>& ingredients, vector<string>& supplies) {
        
        unordered_set<string> all_supplies(supplies.begin(), supplies.end());
        unordered_map<string, vector<string>> recipeGraph;
        unordered_map<string, int> recipe_to_id;
        
        vector<int> waiting_recipe_cnt(recipes.size(), 0);
        queue<int> recipe_ready;
        vector<string> possible_recipes;

        for (int r_id=0; r_id<recipes.size(); r_id++) {

            string recipe = recipes[r_id];
            for (auto ingredient: ingredients[r_id])
                if (!all_supplies.count(ingredient)) {

                    recipeGraph[ingredient].push_back(recipe);
                    waiting_recipe_cnt[r_id]++;

                }   
            
            if (!waiting_recipe_cnt[r_id]) recipe_ready.push(r_id);
            recipe_to_id[recipe] = r_id;

        }

        while (!recipe_ready.empty()) {

            int r_id = recipe_ready.front();
            recipe_ready.pop();
            string recipe = recipes[r_id];
            possible_recipes.push_back(recipe);

            if (!recipeGraph.count(recipe)) continue;

            for (string waiting_recipe: recipeGraph[recipe]) {

                int waiting_r_id = recipe_to_id[waiting_recipe];
                if (!(--waiting_recipe_cnt[waiting_r_id]))
                    recipe_ready.push(waiting_r_id);

            }

        }

        return possible_recipes;

    }
};

File: Problems/No.2115_Find_All_Possible_Recipes_from_Given_Supplies/find_all_possible_recipes_from_given_supplies.py
class Solution(object):
    def findAllRecipes(self, recipes, ingredients, supplies):
        """
        :type recipes: List[str]
        :type ingredients: List[List[str]]
        :type supplies: List[str]
        :rtype: List[str]
        """
        
        all_supplies = set(supplies)
        recipeGraph = defaultdict(list)
        waiting_recipe_cnt = [0]*len(recipes)
        recipe_ready = []
        possible_recipes = []

        for r_id, recipe in enumerate(recipes):
            for ingredient in ingredients[r_id]:
                if ingredient not in all_supplies:
                    recipeGraph[ingredient].append([r_id, recipe])
                    waiting_recipe_cnt[r_id] += 1
            if waiting_recipe_cnt[r_id] == 0:
                recipe_ready.append([r_id, recipe])

        while recipe_ready:
            r_id, recipe = recipe_ready.pop(0)
            possible_recipes.append(recipe)

            if recipe not in recipeGraph:
                continue
            
            for waiting_r_id, waiting_recipe in recipeGraph[recipe]:
                waiting_recipe_cnt[waiting_r_id] -= 1
                if not waiting_recipe_cnt[waiting_r_id]:
                    recipe_ready.append([waiting_r_id, waiting_recipe])

        return possible_recipes
